libraries/ts-command-line/src/providers/TabCompletionAction.ts

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import stringArgv from 'string-argv';
+
+import { CommandLineIntegerParameter } from '../parameters/CommandLineIntegerParameter';
+import { CommandLineStringParameter } from '../parameters/CommandLineStringParameter';
+import { CommandLineParameterKind, CommandLineParameter } from '../parameters/BaseClasses';
+import { CommandLineAction } from './CommandLineAction';
+import { CommandLineChoiceParameter } from '..';
+import { CommandLineConstants } from '../Constants';
+
+const DEFAULT_WORD_TO_AUTOCOMPLETE: string = '';
+const DEFAULT_POSITION: number = 0;
+
+export class TabCompleteAction extends CommandLineAction {
+  private _wordToCompleteParameter: CommandLineStringParameter;
+  private _positionParameter: CommandLineIntegerParameter;
+  private readonly _actions: Map<string, Map<string, CommandLineParameter>>;
+  private readonly _globalParameters: Map<string, CommandLineParameter>;
+
+  public constructor(
+    actions: ReadonlyArray<CommandLineAction>,
+    globalParameters: ReadonlyArray<CommandLineParameter>
+  ) {
+    super({
+      actionName: CommandLineConstants.TabCompletionActionName,
+      summary: 'Provides tab completion.',
+      documentation: 'Provides tab completion.'
+    });
+
+    this._actions = new Map<string, Map<string, CommandLineParameter>>();
+    for (const action of actions) {
+      const parameterNameToParameterInfoMap: Map<string, CommandLineParameter> = new Map<
+        string,
+        CommandLineParameter
+      >();
+      for (const parameter of action.parameters) {
+        parameterNameToParameterInfoMap.set(parameter.longName, parameter);
+        if (parameter.shortName) {
+          parameterNameToParameterInfoMap.set(parameter.shortName, parameter);
+        }
+      }
+      this._actions.set(action.actionName, parameterNameToParameterInfoMap);
+    }
+
+    this._globalParameters = new Map<string, CommandLineParameter>();
+    for (const parameter of globalParameters) {
+      this._globalParameters.set(parameter.longName, parameter);
+      if (parameter.shortName) {
+        this._globalParameters.set(parameter.shortName, parameter);
+      }
+    }
+  }
+
+  protected onDefineParameters(): void {
+    this._wordToCompleteParameter = this.defineStringParameter({
+      parameterLongName: '--word',
+      argumentName: 'WORD',
+      description: `The word to complete.`,
+      defaultValue: DEFAULT_WORD_TO_AUTOCOMPLETE
+    });
+
+    this._positionParameter = this.defineIntegerParameter({
+      parameterLongName: '--position',
+      argumentName: 'INDEX',
+      description: `The position in the word to be completed.`,
```

Comment:
How would this be used?  If it's a complicated explanation, maybe you could provide the URL to some article about POSIX command line completion.

Reply:
It's basically the zero-based-index in the command line string the cursor currently is at when `tab` is pressed. For example, in `rush <tab>`, `--position` will be 5.  In `rush i<tab>`, `--position` will be 6.

Reply:
Hmmm... Then if `--position` is omitted, should it default to `word.length` instead of 0?

Reply:
Good point.